launcher/agent/agent_test.go

```diff
@@ -338,6 +344,170 @@ func TestFetchContainerImageSignatures(t *testing.T) {
+	return r.result, r.err
+}
+
+func intMin(a, b int) int {
+	return int(math.Min(float64(a), float64(b)))
+}
+
+func TestFetchContainerImageSignatures_RetriesOnFailure(t *testing.T) {
+	ctx := context.Background()
+
+	testCases := []struct {
+		name      string
+		resultmap map[string][]returnVal
+	}{
+		{
+			name: "one repo, no failures",
+			resultmap: map[string][]returnVal{
+				"repo1": {
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+			},
+		},
+		{
+			name: "one repo fails",
+			resultmap: map[string][]returnVal{
+				"repo1": {
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data", oci.ECDSAP256SHA256)},
+						err:    fmt.Errorf("partial error"),
+					},
+				},
+			},
+		},
+		{
+			name: "one repo, failure then success",
+			resultmap: map[string][]returnVal{
+				"repo1": {
+					returnVal{
+						result: []oci.Signature{},
+						err:    fmt.Errorf("failure 1"),
+					},
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+			},
+		},
+		{
+			name: "two repos, no failures",
+			resultmap: map[string][]returnVal{
+				"repo1": {
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+				"repo2": {
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data again", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+			},
+		},
+		{
+			name: "two repos, failure then success",
+			resultmap: map[string][]returnVal{
+				"failrepo": {
+					returnVal{
+						result: []oci.Signature{},
+						err:    fmt.Errorf("failure 1"),
+					},
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+				"successRepo": {
+					returnVal{
+						result: []oci.Signature{cosign.NewFakeSignature("test data again", oci.ECDSAP256SHA256)},
+						err:    nil,
+					},
+				},
+			},
+		},
+		{
+			name: "two repos, failures",
+			resultmap: map[string][]returnVal{
+				"repo1": {
+					returnVal{
+						result: []oci.Signature{},
+						err:    fmt.Errorf("failure 1"),
+					},
+				},
+				"repo2": {
+					returnVal{
+						result: []oci.Signature{},
+						err:    fmt.Errorf("failure 2"),
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tpm := test.GetTPM(t)
+			defer client.CheckedClose(t, tpm)
+
+			sdClient := NewFailingClient(tc.resultmap)
+			b := backoff.NewConstantBackOff(time.Millisecond)
+
+			repos := []string{}
+			wantSigs := []oci.Signature{}
+			for k, v := range tc.resultmap {
+				repos = append(repos, k)
+				for _, result := range v {
+					if result.err == nil {
+						wantSigs = append(wantSigs, result.result...)
+					}
+				}
+			}
```

Comment:
I think you're right for table driven tests, will make the change.
I wish table driven tests were different :(

Reply:
I will handle this in a follow up (maybe)
I think I prefer this style of not having to copy a bunch of strings in the input.

Since I want to move things around here anyway, lets talk about this in our 1 on 1.
You are almost certainly right because this is what the tests look like now, but lets chat.,